text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+`aws-cdk-lib` by approximately 130 MiB*, and make the Lambda Layers the biggest
+contributor of size at 39%.
+
+*See Appendix A to see where this estimate came from.
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| Lambda Layer zip files         | 38.83%                         |
+| .jsii files                    | 30.14%                         |
+| Source Maps (.js.map)          | 15.87%                         |
+| Javascript code (.js)          | 6.63%                          |
+| Type Declaration files (.d.ts) | 6.62%                          |
+| README, etc (.md)              | 0.55%                          |
+| bundled npm dependencies       | 0.50%                          |
+| .ts-fixture                    | 0.27%                          |
+| .jsiirc                        | 0.26%                          |
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
+
+### What is the technical solution (design) of this feature?
+
+There are different categories of files contributing to the size of aws-cdk-lib,
+and each category will have it’s own solution. The following sections cover the
+top five categories in size contribution in order from largest to smallest
+potential size.
+
+#### Lambda Layer Zip Files
+
+Use separate npm packages for each `lambda-layer-X` module, which the CDK CLI
+will automatically install if not already present in the `node_modules`
+directory. With this solution, customers will not have to figure out when they
+do or do not need to include these dependencies, and they will not be required
+to access a public endpoint besides npm during runtime of their CDK apps.
+Practically, this solution breaks down into the following steps.
+
+1. Publish v2 compatible versions of `@aws-cdk/lambda-layer-aws-cli`,
+   `@aws-cdk/lambda-layer-aws-cli`, `@aws-cdk/lambda-layer-node-proxy-agent` as
+   their own npm packages, separate from `aws-cdk-lib`. This publishing process
+   will be an extension of the existing AWS CDK V2 release machinery. Any
+   submodules with the prefix `lambda-layer` will be published the same way.
+2. Modify the `lambda-layer-X` submodules in `aws-cdk-lib` to not bundle these
+   large dependencies themselves, but instead reference the appropriate `Layer`
+   class from the packages in step 1. Since `aws-cdk-lib`'s source code is
+   auto-generated from the v1 submodules, this could be an autogenerated
+   wrapper. We will need to do something clever here to make sure customers do
+   not have IDE or compilation errors before the CDK CLI has an opportunity to
+   download and install the packages.
```

Comment:
If the external NPM package is just the binary (with some wrapping JS to interface with), this would be trivial.

Reply:
Yeah you are right. This package will be the `Layer` construct including the binary, but it will still be an implementation detail of the constructs exposed in `aws-cdk-lib`. Updated this section.